File: NeetCode/Trie/Implement_Trie_Prefix_Tree/implement__trie__prefix__tree.py
from typing import Optional, List
import string


class Node:
    def __init__(self):
        self.children = {}
        self.terminal = False


class PrefixTree:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        ptr = self.root
        for w in word:
            if w not in ptr.children:
                ptr.children[w] = Node()
            ptr = ptr.children[w]
        ptr.terminal = True

    def search(self, word: str) -> bool:
        ptr = self.root
        for w in word:
            if w not in ptr.children:
                return False
            ptr = ptr.children[w]
        return ptr.terminal

    def startsWith(self, prefix: str) -> bool:
        ptr = self.root
        for w in prefix:
            if w not in ptr.children:
                return False
            ptr = ptr.children[w]
        return ptr != None


if __name__ == "__main__":
    prefixTree = PrefixTree()
    print(prefixTree.insert("dog"))
    print(prefixTree.search("dog"))
    print(prefixTree.search("do"))
    print(prefixTree.startsWith("do"))
    print(prefixTree.insert("do"))
    print(prefixTree.search("do"))
    print("Running Solution...")

File: NeetCode/Trie/Implement_Trie_Prefix_Tree/Implement_Trie_Prefix_Tree.cpp
// Implement_Trie_Prefix_Tree.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class PrefixTree
{
private:
    class Node
    {
    public:
        array<Node *, 26> children{};
        bool terminal = false;
    };
    Node *root;

public:
    PrefixTree()
    {
        root = new Node();
    }

    void insert(string word)
    {
        Node *ptr = root;
        for (char ch : word)
        {
            if (ptr->children[ch - 'a'] == nullptr)
            {
                ptr->children[ch - 'a'] = new Node();
            }
            ptr = ptr->children[ch - 'a'];
        }
        ptr->terminal = true;
    }

    bool search(string word)
    {
        Node *ptr = root;
        for (char ch : word)
        {
            if (ptr->children[ch - 'a'] == nullptr)
            {
                return false;
            }
            ptr = ptr->children[ch - 'a'];
        }
        return ptr->terminal;
    }

    bool startsWith(string prefix)
    {
        Node *ptr = root;
        for (char ch : prefix)
        {
            if (ptr->children[ch - 'a'] == nullptr)
            {
                return false;
            }
            ptr = ptr->children[ch - 'a'];
        }
        return true;
    }
};
